auth/src/main/java/com/redhat/lightblue/rest/auth/jboss/CertLdapLoginModule.java

```diff
@@ -99,49 +99,51 @@ public void initialize(Subject subject, CallbackHandler callbackHandler, Map<Str
 
     public void initialize() throws Exception {
 
-        environment = (String) options.get(ENVIRONMENT);
-
-        LdapConfiguration ldapConf = new LdapConfiguration();
-
-        ldapConf.server((String) options.get(SERVER));
-        ldapConf.port(Integer.parseInt((String) options.get(PORT)));
-        String searchBase = (String) options.get(SEARCH_BASE);
-        ldapConf.bindDn((String) options.get(BIND_DN));
-        ldapConf.bindDNPwd((String) options.get(BIND_PWD));
-        ldapConf.useSSL(Boolean.parseBoolean( (String) options.get(USE_SSL)));
-        ldapConf.trustStore((String) options.get(TRUST_STORE));
-        ldapConf.trustStorePassword((String) options.get(TRUST_STORE_PASSWORD));
-        ldapConf.poolSize(Integer.parseInt((String) options.get(POOL_SIZE)));
-
-        // optional configurations
-        if (options.containsKey(CONNECTION_TIMEOUT_MS)) {
-            ldapConf.connectionTimeoutMS(Integer.parseInt((String)options.get(CONNECTION_TIMEOUT_MS)));
-        }
-        if (options.containsKey(RESPONSE_TIMEOUT_MS)) {
-            ldapConf.responseTimeoutMS(Integer.parseInt((String)options.get(RESPONSE_TIMEOUT_MS)));
-        }
-        if (options.containsKey(DEBUG)) {
-            ldapConf.debug(Boolean.parseBoolean((String)options.get(DEBUG)));
-        }
-        if (options.containsKey(KEEP_ALIVE)) {
-            ldapConf.keepAlive(Boolean.parseBoolean((String)options.get(KEEP_ALIVE)));
-        }
-        if (options.containsKey(POOL_MAX_CONNECTION_AGE_MS)) {
-            ldapConf.poolMaxConnectionAgeMS(Integer.parseInt((String)options.get(POOL_MAX_CONNECTION_AGE_MS)));
-        }
-
-        int rolesCacheExpiry = 5*60*1000; // default 5 minutes
-        if (options.containsKey(ROLES_CACHE_EXPIRY_MS)) {
-            rolesCacheExpiry = Integer.parseInt((String)options.get(ROLES_CACHE_EXPIRY_MS));
-        }
+        synchronized(LdapRolesProvider.class) {
```

Comment:
> Volatile doesn't mean what you think, either
> A commonly suggested nonfix is to declare the resource field of SomeClass as volatile. However, while the JMM prevents writes to volatile variables from being reordered with respect to one another and ensures that they are flushed to main memory immediately, it still permits reads and writes of volatile variables to be reordered with respect to nonvolatile reads and writes. That means -- unless all Resource fields are volatile as well -- thread B can still perceive the constructor's effect as happening after resource is set to reference the newly created Resource.

(from http://www.javaworld.com/article/2074979/java-concurrency/double-checked-locking--clever--but-broken.html)

If I understand this correctly (and unless something has changed, it's a really old article), volatile will not affect the properties of LdapRolesProvider, unless they are volatile as well.

There is also https://en.wikipedia.org/wiki/Initialization-on-demand_holder_idiom, but it only works for initialization which cannot fail, so not good.